editor/src/utils/react-performance.ts

```diff
@@ -209,141 +209,188 @@ function failSafeMemoEqualityFunction(componentDisplayName: string, severity: 's
   }
 }
 
+// this function has been adopted from https://github.com/epoberezkin/fast-deep-equal/tree/a33d49ab5cc659e331ff445109f35dd323230d41
 function keepDeepReferenceEqualityInner(
-  oldValueInner: any,
-  possibleNewValueInner: any,
+  oldValue: any,
+  possibleNewValue: any,
   stackSizeInner: number,
-  valueStackSoFar: Array<any>,
-): any {
-  // most of the actual structure is copy-pasted from https://github.com/epoberezkin/fast-deep-equal/blob/v2/index.js
-  // I've made it return oldValue or newValue instead of true and false
-  // and the recursion only stops if we get to return oldValue
-  // otherwise we still drill into non-equal objects to try and find
-  // keys that can be made equal
-
-  // If we're more than 100 frames deep, let's just call it a day
+  valueStackSoFar: Set<any>,
```

Comment:
`Set<any>` scares me but also it's potentially not a stack of values if stored in a `Set` because it's not really designed to keep things in order.

Reply:
I couldn't come up with a better name. I guess `valueSetSoFar` would be uhmm maybe better?

I changed it to Set because `Set.has` is much faster than `Array.includes` according to some googling, and it sounds commmon sense to me because Array.includes must iterate over the array

https://github.com/concrete-utopia/utopia/pull/657

Reply:
Surely if `Set` is just using `===` it must also be iterating over some internal array? Unless there's some internal magic optimisation that we're not privy to.

Reply:
I believe it to be optimized, but I don't know the details.

https://dev.to/arnaud/using-array-prototype-includes-vs-set-prototype-has-to-filter-arrays-41fg

Reply:
It would be worth confirming if that holds for things other than primitive values (and or those containers with arbitrary types in them). This has the whiff of `Set` using hashed buckets to hold the values which is super easy for numbers as you can just use the number as a hash.

Reply:
good point, I'm checking it! I'll post a CSB here

Reply:
https://codesandbox.io/s/arrayincludes-vs-sethas-lr63q?file=/src/index.ts

Reply:
okay this is actually a wow result. with a very large array (100000 elements) the difference is 1000x!

```
Array.includes for 1000 elements took 3.1350000062957406/1000 ms 
Set.has for 1000 elements took 0.5050000036135316/1000 ms 
Array.includes for 10000 elements took 26.700000002165325/1000 ms 
Set.has for 10000 elements took 0.6600000051548705/1000 ms 
Array.includes for 100000 elements took 260.9450000018114/1000 ms 
Set.has for 100000 elements took 0.28500000189524144/1000 ms 
```

It seems like set has an O(1) lookup time. For use cases like this, when we want to maintain a short lived set of elements for filtering out stuff, Set's downsides (=== equality of elements, cannot serialize, etc) doesn't seem to be problematic at all. and the upside is _huge_.